shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm

```diff
@@ -24,124 +44,129 @@
 #include "flutter/fml/task_runner.h"
 #include "flutter/runtime/dart_service_isolate.h"
 
-#if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_RELEASE || \
-    FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DYNAMIC_RELEASE
+@protocol FlutterObservatoryPublisherDelegate
+- (instancetype)initWithOwner:(FlutterObservatoryPublisher*)owner;
+- (void)publishServiceProtocolPort:(NSString*)uri;
+- (void)stopService;
 
-@implementation FlutterObservatoryPublisher {
-}
+@property(readonly) fml::scoped_nsobject<NSURL> url;
+@end
+
+@interface FlutterObservatoryPublisher ()
+- (NSData*)createTxtData:(NSURL*)url;
 
-#else
+@property(readonly) NSString* serviceName;
+@property(readonly) fml::scoped_nsobject<NSObject<FlutterObservatoryPublisherDelegate>> delegate;
 
-@interface FlutterObservatoryPublisher () <NSNetServiceDelegate>
 @end
 
-@implementation FlutterObservatoryPublisher {
-  fml::scoped_nsobject<NSURL> _url;
-#if TARGET_IPHONE_SIMULATOR
-  DNSServiceRef _dnsServiceRef;
-#else   // TARGET_IPHONE_SIMULATOR
-  fml::scoped_nsobject<NSNetService> _netService;
-#endif  // TARGET_IPHONE_SIMULATOR
+@interface ObservatoryNSNetServiceDelegate
+    : NSObject <FlutterObservatoryPublisherDelegate, NSNetServiceDelegate>
+@end
 
-  flutter::DartServiceIsolate::CallbackHandle _callbackHandle;
-  std::unique_ptr<fml::WeakPtrFactory<FlutterObservatoryPublisher>> _weakFactory;
-}
+@interface ObservatoryDNSServiceDelegate : NSObject <FlutterObservatoryPublisherDelegate>
+@end
 
-- (NSURL*)url {
-  return _url.get();
+@implementation ObservatoryDNSServiceDelegate {
+  fml::scoped_nsobject<FlutterObservatoryPublisher> _owner;
+  DNSServiceRef _dnsServiceRef;
 }
 
-- (instancetype)init {
+@synthesize url;
```

Comment:
Do you need this `@synthesize`?

Reply:
AFAIK yes - because it's from a protocol and not autosynthesized.  At any rate, doesn't compile without this.

Reply:
Oh that make sense.

Reply:
I think this needs an ivar, how is this not crashing when you use url?
```objective-c
@implementation ObservatoryDNSServiceDelegate {
  fml::scoped_nsobject<FlutterObservatoryPublisher> _owner;
  DNSServiceRef _dnsServiceRef;
  fml::scoped_nsobject<NSURL> _url;
}

@synthesize url = _url;
```
```objective-c
self.url.reset([[[NSURL alloc] initWithString:uri] retain]);
```

Reply:
From our conversation offline - this doesn't work.

I think this is because fml::scoped_nsobject is a stack allocated C++ object, rather than a pointer or a NSObject.  It makes things look really funky compared to how this would be done if we were using ARC  (where the property/ivar would just be something like a raw `NSURL*`). But this syntax compiles and runs.